### ISSUE TITLE
Uninstalled app is not redeployed

### ISSUE THREAD

**Issue by jpobst** [xamarin/xamarin-android#3727]
### Steps to Reproduce

1. F5 on app using Shared Runtime + Fast Deployment
2. App runs fine
3. Close app
4. Uninstall app on device
5. F5 

### Expected Behavior

App is redeployed to device and debugging succeeds.

### Actual Behavior

```
Error: Device could not find component named: com.companyname.deleteable/md52d2c55cfc97a76782f55285793be350d.MainActivity
The application could not be started. Ensure that the application has been installed to the target device and has a launchable activity (MainLauncher = true).
Additionally, check Build->Configuration Manager to ensure this project is set to Deploy for this configuration.
Couldn't connect debugger. You can see more details in Xamarin Diagnostic output and the full exception on logs.
```

Note that retrying a second time succeeds.

### Version Information

VS 2019 16.3.2

### Log File

[build.log](https://github.com/xamarin/xamarin-android/files/3682121/build.log)

**Comment by dellis1972** [comment]
So it looks like out Task which detects if the app has been uninstalled is not completing before the `_Upload` target runs. So we need to take a look at that. 

Also we should check to make sure the Shared Runtime etc is still installed.

**Comment by dellis1972** [comment]
Hmm we already have a unit test for this https://github.com/xamarin/xamarin-android/blob/master/tests/MSBuildDeviceIntegration/Tests/InstallTests.cs#L35

**Comment by dellis1972** [comment]
A fix for this issue was merged via xamarin/monodroid@663fa21  and  xamarin/monodroid@400af80

**Comment by brendanzagaeski** [comment]
*Release status update*

A new Preview version has now been published on Windows that includes the fix for this item. The fix is not yet included in a Release version. I will update this item again when a Release version is available that includes the fix.

The fix is not yet available on macOS. I will update this item again when a Preview version with the fix is available on macOS.

Fix included in Xamarin.Android 10.2.0.16.

Fix included on Windows in Visual Studio 2019 version 16.5 Preview 1. To try the Preview version that includes the fix, [check for the latest updates](https://docs.microsoft.com/visualstudio/install/update-visual-studio?view=vs-2019) in [Visual Studio Preview](https://visualstudio.microsoft.com/vs/preview/).

Fix not yet available on macOS.

**Comment by brendanzagaeski** [comment]
*Release status update*

A new Preview version has now been published on macOS that includes the fix for this item.

Fix included in Xamarin.Android 10.2.0.84.

Fix included on macOS in Visual Studio 2019 for Mac version 8.5 Preview 1. To try the Preview version that includes the fix, check for the latest updates on the **Preview** [updater channel](https://docs.microsoft.com/visualstudio/mac/update).

(Fix also included on Windows in Visual Studio 2019 version 16.5 Preview 1 and higher. To try the Preview version that includes the fix, [check for the latest updates](https://docs.microsoft.com/visualstudio/install/update-visual-studio?view=vs-2019) in [Visual Studio Preview](https://visualstudio.microsoft.com/vs/preview/).)

**Comment by brendanzagaeski** [comment]
*Release status update*

A new Release version has now been published on Windows that includes the fix for this item. The fix is not yet published in a Release version on macOS. I will update this item again when a Release version is available on macOS that includes the fix.

Fix included in Xamarin.Android 10.2.0.100.

Fix included on Windows in Visual Studio 2019 version 16.5. To get the new version that includes the fix, [check for the latest updates](https://docs.microsoft.com/visualstudio/install/update-visual-studio?view=vs-2019) or install the latest version from <https://visualstudio.microsoft.com/downloads/>.

(Fix also included on macOS in Visual Studio 2019 for Mac version 8.5 Preview 1 and higher. To try the Preview version that includes the fix, check for the latest updates on the **Preview** [updater channel](https://docs.microsoft.com/visualstudio/mac/update).)

**Comment by brendanzagaeski** [comment]
*Release status update*

A new Release version has now been published on macOS that includes the fix for this item.

Fix included in Xamarin.Android 10.2.0.100.

Fix included on macOS in Visual Studio 2019 for Mac version 8.5. To get the new version that includes the fix, check for the latest updates on the **Stable** [updater channel](https://docs.microsoft.com/visualstudio/mac/update).

(Fix also included on Windows in Visual Studio 2019 version 16.5 and higher. To get the new version that includes the fix, [check for the latest updates](https://docs.microsoft.com/visualstudio/install/update-visual-studio?view=vs-2019) or install the latest version from <https://visualstudio.microsoft.com/downloads/>.)

**Comment by EJocys** [comment]
I still have same bug. Visual Studio & Xamarin just frequently fails to deploy and fails to start debugging session with:

Error: Device could not find component named: com.package.name/crc&lt;MD5&gt;.SplashActivity
The application could not be started. Ensure that the application has been installed to the target device and has a launchable activity (MainLauncher = true).
Additionally, check Build->Configuration Manager to ensure this project is set to Deploy for this configuration.
Couldn't connect debugger. You can see more details in Xamarin Diagnostic output and the full exception on logs.

So, I must to use reliable workaround, which is:

Forcing Visual Studio to create *.apk files on every debug build by adding element inside Debug configuration section of  *.Android.csproj file:

```xml
<AndroidBuildApplicationPackage>true</AndroidBuildApplicationPackage>
```

Running DOS batch script ClearApp_Reinstall.bat, which will deploys new debug *.apk manually to the Android phone or emulator:
```bash
SET app=com.package.name
SET adb=%ProgramFiles(x86)%\Android\android-sdk\platform-tools\adb.exe
"%adb%" uninstall "%app%"
"%adb%" push "Mobile.Android\bin\Debug\%app%-Signed.apk" "/data/local/tmp"
"%adb%" shell chmod 750 "/data/local/tmp/*.apk"
"%adb%" shell pm install -t -r "/data/local/tmp/%app%-Signed.apk"
```

Basically, I build, run the batch file and then start debugging.

**Comment by dellis1972** [comment]
@EJocys if you are getting this on a regular basis could you install the `Project System Tools` extension in Visual Studio (https://marketplace.visualstudio.com/items?itemName=VisualStudioProductTeam.ProjectSystemTools). 

You can then use that to get us the `binlog` files we need to figure out what is happening on your machine. You can see the instructions on how to use the tool here https://github.com/dotnet/project-system-tools#features. 

Basically we need the logs for the `Build` and `Install` targets, they will appear in the `Project System Tools` window. 

Under normal operations what should happen is you click Run in Visual Studio and it will detect if the app a) has been updated and needs to be installed or b) does not exist on the device and needs to be installed. The `Install` target is the msbuild target which handles that. It is called by the IDE when you hit run. 

It would also be interesting to see the settings you have in your Debug configuration so we can make sure you are correctly setup for debugging.

**Comment by EJocys** [comment]
Issue starts with Visual Studio failing to build and deploy *.apk, which contains changed code. VS just just skips deployment process and starts old app for debugging. So, as a workaround I am uninstalling the app directly from the phone. This will trigger "Error: Device could not find component named:..." error when trying to debug. Visual Studio throws this error immediately, without any building or deploying. Build logging shows event below and trying to save anything saves nothing:

![FailingToBuild](https://user-images.githubusercontent.com/11958537/80494128-38237380-895e-11ea-891f-de80575c3aa8.png)

Now there is interesting part. If I open Android project properties, go to "Android Options" and keep tab open (which makes Visual Studio not to loose focus from the Android Project in Solution Explorer) then sooner or later this line appears:

![FailingToBuild_2](https://user-images.githubusercontent.com/11958537/80494950-2e4e4000-895f-11ea-90b6-6d41f1a99cb0.png)

Now, it is possible to save some logs. Also, selecting Android project and doing "Debug -> Start New Instance" results in Visual Studio deploying and starting new debug session successfully.

So, opening "Android Options" will trigger some event, which will resolve the error which is the topic of this issue.

**Comment by dellis1972** [comment]
@EJocys changing C# code should not need to re-install the apk if you are using the fast deployment system which should be on by default. You should not need to keep uninstalling the apk each time. Can you share the debug configuration for your application? I suspect you have some settings turned off/on which are interfering with the way our fasted system works. 

As for the logs, you need to keep an eye out for entries which have `Build` or `Install` in the targets column. Not sure why its not saving data.

**Comment by EJocys** [comment]
@dellis1972:  I would be glad to use fast deployment but it was just not working reliably. I would end up debugging old code, because Visual Studio would fail to deploy new code. I've started to use reliable alternatives exactly because Visual Studio auto-deployment options were not reliable. 

Visual Studio error suggests that it can't find component "com.package.name/md&gt;MD5&lt;.MainActivity" when all "com.package.name" app is missing. It would be more reasonable error if it said:
"Error: could not find application named: com.package.name"

Debug config with sensitive info removed. It can create *.apk if forced no problem. Deployment is the issue:

```
  <PropertyGroup>
    <Configuration Condition=" '$(Configuration)' == '' ">Debug</Configuration>
    <Platform Condition=" '$(Platform)' == '' ">AnyCPU</Platform>
    <ProjectGuid>{********-****-****-****-************}</ProjectGuid>
    <ProjectTypeGuids>{********-****-****-****-************};{********-****-****-****-************}</ProjectTypeGuids>
    <TemplateGuid>{********-****-****-****-************}</TemplateGuid>
    <OutputType>Library</OutputType>
    <RootNamespace>*****.Mobile.Droid</RootNamespace>
    <AssemblyName>*****.Mobile.Android</AssemblyName>
    <AndroidApplication>True</AndroidApplication>
    <AndroidResgenFile>Resources\Resource.designer.cs</AndroidResgenFile>
    <AndroidResgenClass>Resource</AndroidResgenClass>
    <AndroidManifest>Properties\AndroidManifest.xml</AndroidManifest>
    <MonoAndroidResourcePrefix>Resources</MonoAndroidResourcePrefix>
    <MonoAndroidAssetsPrefix>Assets</MonoAndroidAssetsPrefix>
    <AndroidUseLatestPlatformSdk>false</AndroidUseLatestPlatformSdk>
    <TargetFrameworkVersion>v9.0</TargetFrameworkVersion>
    <AndroidEnableSGenConcurrent>true</AndroidEnableSGenConcurrent>
    <AndroidUseAapt2>true</AndroidUseAapt2>
    <AndroidHttpClientHandlerType>Xamarin.Android.Net.AndroidClientHandler</AndroidHttpClientHandlerType>
    <NuGetPackageImportStamp>
    </NuGetPackageImportStamp>
    <SccProjectName>SAK</SccProjectName>
    <SccLocalPath>SAK</SccLocalPath>
    <SccAuxPath>SAK</SccAuxPath>
    <SccProvider>SAK</SccProvider>
  </PropertyGroup>
  <PropertyGroup Condition=" '$(Configuration)|$(Platform)' == 'Debug|AnyCPU' ">
    <DebugSymbols>true</DebugSymbols>
    <DebugType>portable</DebugType>
    <Optimize>false</Optimize>
    <OutputPath>bin\Debug</OutputPath>
    <DefineConstants>DEBUG;</DefineConstants>
    <ErrorReport>prompt</ErrorReport>
    <WarningLevel>4</WarningLevel>
    <AndroidLinkMode>None</AndroidLinkMode>
    <AotAssemblies>false</AotAssemblies>
    <EnableLLVM>false</EnableLLVM>
    <AndroidEnableProfiledAot>false</AndroidEnableProfiledAot>
    <BundleAssemblies>false</BundleAssemblies>
    <AndroidKeyStore>true</AndroidKeyStore>
    <AndroidSigningKeyStore>$(SolutionDir)\Resources\Keystore\*****.mobile.keystore</AndroidSigningKeyStore>
    <AndroidSigningKeyAlias>*****.mobile</AndroidSigningKeyAlias>
    <AndroidSigningKeyPass>*****</AndroidSigningKeyPass>
    <AndroidSigningStorePass>*****</AndroidSigningStorePass>
    <AndroidUseSharedRuntime>true</AndroidUseSharedRuntime>
    <EmbedAssembliesIntoApk>true</EmbedAssembliesIntoApk>
    <AndroidSupportedAbis>armeabi-v7a;x86;x86_64;arm64-v8a</AndroidSupportedAbis>
    <MandroidI18n />
    <AndroidUseAapt2>true</AndroidUseAapt2>
    <AndroidHttpClientHandlerType>System.Net.Http.HttpClientHandler</AndroidHttpClientHandlerType>
    <AndroidCreatePackagePerAbi>false</AndroidCreatePackagePerAbi>
    <AndroidPackageFormat>apk</AndroidPackageFormat>
  </PropertyGroup>
... 
<PropertyGroup>
    <PostBuildEvent>echo Build Log: $(IntDir)\$(MSBuildProjectName).log</PostBuildEvent>
  </PropertyGroup>
<

```
